File: src/main/scala/com/weave/core/Reducer.scala
package com.weave.core

import org.apache.pekko.util.Collections

import scala.collection.immutable.{AbstractSeq, LinearSeq}

trait Reducer[S, U] {
  def reduce(state: S, update: U): S
  def merge(left: S, right: S): S
  def merge(first: S, states: S*): S = states.foldLeft(first)(merge)
}

File: kotlin-reference/src/main/kotlin/com/weave/core/Reducer.kt
package com.weave.core

public interface Reducer<S, U> {
    public fun reduce(state: S, update: U): S

    public fun merge(left: S, right: S): S

    public fun merge(first: S, vararg states: S): S {
        var result = first
        states.forEach { state -> result = merge(result, state) }
        return result
    }
}
